vimiv/imutils/metadata.py

```diff
@@ -0,0 +1,284 @@
+`MetadataHandler` is used to interact with the metadata of the current image. It
+relies on the functionality provided by optionally loaded metadata plugins. Such
+plugins implements the `MetadataPlugin` abstract class and registers that class
+using the `register` function.
+
+Module Attributes:
+    _registry: List of registered `MetadataPlugin` implementations.
+"""
+
+from abc import ABC, abstractmethod
+import contextlib
+import itertools
+from typing import Dict, Tuple, NoReturn, Sequence, Iterable, Type, List
+
+from vimiv.utils import log
+
+_logger = log.module_logger(__name__)
+
+# Type returned by `MetadataHandler.get_metadata`.
+# Key is the metadata key. Value is a tuple of descriptive name and value for that key.
+MetadataDictT = Dict[str, Tuple[str, str]]
+
+
+class MetadataPlugin(ABC):
+    """Abstract class implemented by plugins to provide metadata capabilities.
+
+    Implementations of this class are required to overwrite `__init__`, `name`,
+    `version`, `get_metadata` and `get_keys`.
+    The implementations of `copy_metadata` and `get_date_time` is optional.
+    """
+
+    @abstractmethod
+    def __init__(self, _path: str) -> None:
+        """Initialize metadata handler for a specific image.
+
+        Args:
+            _path: Path to current image.
+        """
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """Get the name of the used backend.
+
+        If no backend is used, return the name of the plugin.
+        """
+
+    @property
+    @abstractmethod
+    def version(self) -> str:
+        """Get the version of the used backend.
+
+        If no backend is used, return an empty string.
+        """
+
+    @abstractmethod
+    def get_metadata(self, _keys: Sequence[str]) -> MetadataDictT:
+        """Get value of all desired keys for the current image.
+
+        If no value is found for a certain key, do not include the key in the output.
+
+        Args:
+            _keys: Keys of metadata to query the image for.
+
+        Returns:
+            Dictionary with retrieved metadata.
+        """
+
+    @abstractmethod
+    def get_keys(self) -> Iterable[str]:
+        """Get the keys for all metadata values available for the current image."""
+
+    def copy_metadata(self, _dest: str, _reset_orientation: bool = True) -> bool:
+        """Copy metadata from the current image to dest image.
+
+        Args:
+            _dest: Path to write the metadata to.
+            _reset_orientation: If true, reset the exif orientation tag to normal.
+
+        Returns:
+            Flag indicating if copy was successful.
+        """
+        raise NotImplementedError
+
+    def get_date_time(self) -> str:
+        """Get creation date and time of the current image as formatted string."""
+        raise NotImplementedError
+
+
+# Stores all registered metadata implementations.
+_registry: List[Type[MetadataPlugin]] = []
```

Comment:
Type annotation should be `List[MetadataPlugin]`, no?

Reply:
No, because plugins call `metadata.register` by passing the implementation of `MetadadaPlugin` as the argument (i.e. the class itself). And while the type of instances of the class `MetadataPlugin` is `MetadadaPlugin`, the class itself is of type `Type[MetadadaPlugin]`.

Reply:
Sorry for the noise then :blush: